contracts/token/ERC721/ERC721.sol

```diff
@@ -343,41 +312,51 @@ abstract contract ERC721 is Context, ERC165, IERC721, IERC721Metadata, IERC721Er
      *
      * Emits a {Transfer} event.
      */
-    function _transfer(address from, address to, uint256 tokenId) internal virtual {
-        address owner = ownerOf(tokenId);
-        if (owner != from) {
-            revert ERC721IncorrectOwner(from, tokenId, owner);
+    function _transfer(address from, address to, uint256 tokenId) internal {
+        if (from == address(0)) {
+            revert ERC721InvalidSender(address(0));
         }
         if (to == address(0)) {
             revert ERC721InvalidReceiver(address(0));
         }
+        _update(from, to, tokenId);
+    }
 
-        _beforeTokenTransfer(from, to, tokenId, 1);
-
-        // Check that tokenId was not transferred by `_beforeTokenTransfer` hook
-        owner = ownerOf(tokenId);
-        if (owner != from) {
-            revert ERC721IncorrectOwner(from, tokenId, owner);
+    function _update(address from, address to, uint256 tokenId) internal virtual {
+        address owner = _ownerOf(tokenId);
+        if (from != owner) {
+            if (owner == address(0)) {
+                revert ERC721NonexistentToken(tokenId);
+            } else {
+                revert ERC721IncorrectOwner(from, tokenId, owner);
+            }
         }
 
-        // Clear approvals from the previous owner
-        delete _tokenApprovals[tokenId];
-
-        // Decrease balance with checked arithmetic, because an `ownerOf` override may
-        // invalidate the assumption that `_balances[from] >= 1`.
-        _balances[from] -= 1;
+        if (from != address(0)) {
+            delete _tokenApprovals[tokenId];
+            unchecked {
+                _balances[from] -= 1;
+            }
+        }
 
-        unchecked {
-            // `_balances[to]` could overflow in the conditions described in `_mint`. That would require
-            // all 2**256 token ids to be minted, which in practice is impossible.
-            _balances[to] += 1;
+        if (to != address(0)) {
+            unchecked {
+                _balances[to] += 1;
+            }
         }
 
         _owners[tokenId] = to;
 
         emit Transfer(from, to, tokenId);
+    }
 
-        _afterTokenTransfer(from, to, tokenId, 1);
+    /**
+     * @dev Unsafe write access to the balances, used by extensions that "mint" tokens using an {ownerOf} override.
+     */
+    function _updateBalance(address account, uint128 value) internal virtual {
```

Comment:
The natspec is outdated. This function should not be described as unsafe. It will never lead to overflow

Reply:
> I don't understand why updateBalance would sound like it's a getter. 

My bad, I meant it sounds like a **setter** instead, not a value increase.

> increaseBalance is good too, but I thought the parallel with _update made sense.

Makes sense as long as the NatSpec correctly reflects what it does. I changed it for `_increaseBalance` because I'm strongly in favor of that name. We can discuss it.